cpp/src/Ice/ConnectionFactory.cpp

```diff
@@ -1266,26 +1241,8 @@ IceInternal::IncomingConnectionFactory::waitUntilFinished()
 
     {
         lock_guard lock(_mutex);
-        if (_transceiver)
-        {
-            assert(_connections.size() <= 1); // The connection isn't monitored or reaped.
-        }
-        else
-        {
-            // Ensure all the connections are finished and reapable at this point.
-            vector<Ice::ConnectionIPtr> cons;
-            _monitor->swapReapedConnections(cons);
```

Comment:
I don't understand the reaping logic in the ACM code.

Presumably when the new idle check code aborts a connection, this connection gets cleaned up. It doesn't need to be "reaped".

Reply:
The connection doesn't get removed from the factory because the connection doesn't have a reference to the factory. 

When a connection is done, it's added to a "reap connection" collection. This collection is held by the ACM code and it's indeed not correct since it has nothing to do with ACM.

Each time a new connection is added to the outgoing/incoming connection factories, the "reap connection" collection is checked and connections from this collection are removed from the factory.

Removing this code implies that connections will leak until the factory is destroyed.

Either you add another class to hold the "reap connection" collection or you fix the connection class to have a circular reference to the connection factory.

Reply:
It seems simpler for the connection to hold a weak_ptr on the connection factory.

Reply:
Yes, it can probably also just be a shared_ptr given that the connections are closed on factory destruction and the connections maps are cleared in the factories.

Reply:
I tried to remove the connection from the factory in "reap" (see the latest code) but this doesn't work.

For outgoing connections, this results in a lock acquisition order deadlock because reap is always called with the ConnectionI mutex lock. It also hangs for incoming connections but I didn't debug it to identify the issue.

Any suggestion?
I'd rather not start a thread just to cleanup these "finished" connections from time to time in the background.

Reply:
Why do we need to hold the connection lock while calling reap, can we just set a flag and called after we release the connection lock?
Instead of

```
if (completedUpcallCount > 0)
{
    std::lock_guard lock(_mutex);
    ....
        else if (_state == StateFinished)
        {
            reap();
        }
        _conditionVariable.notify_all();
    }
}
```

Do something like:

```
bool reapConnection;
if (completedUpcallCount > 0)
{
    std::lock_guard lock(_mutex);
    ....
        else if (_state == StateFinished)
        {
            reapConnection = true;
        }
        _conditionVariable.notify_all();
    }
}

if (reapConnection)
{
    reap();
}
````